src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs

```diff
@@ -67,434 +113,658 @@ public JsonWebToken(string header, string payload)
         /// </summary>
-        /// <remarks>If the 'nbf' claim is not found, then <see cref="DateTime.MinValue"/> is returned.</remarks>
-        public override DateTime ValidFrom => JwtTokenUtilities.GetDateTime(JwtRegisteredClaimNames.Nbf, Payload);
+        public bool IsSigned { get; internal set; }
 
         /// <summary>
-        /// Gets the 'value' of the 'exp' claim { exp, 'value' } converted to a <see cref="DateTime"/> assuming 'value' is seconds since UnixEpoch (UTC 1970-01-01T0:0:0Z).
+        ///
         /// </summary>
-        /// <remarks>If the 'exp' claim is not found, then <see cref="DateTime.MinValue"/> is returned.</remarks>
-        public override DateTime ValidTo => JwtTokenUtilities.GetDateTime(JwtRegisteredClaimNames.Exp, Payload);
+        internal JsonClaimSet Payload { get; set; }
 
         /// <summary>
-        /// Gets the 'value' of the 'x5t' claim { x5t, 'value' }.
+        /// Not implemented.
         /// </summary>
-        /// <remarks>If the 'x5t' claim is not found, an empty string is returned.</remarks>   
-        public string X5t => Header.Value<string>(JwtHeaderParameterNames.X5t) ?? string.Empty;
+        public override SecurityKey SecurityKey { get; }
 
         /// <summary>
-        /// Gets the 'value' of the 'zip' claim { zip, 'value' }.
+        /// Gets or sets the <see cref="SecurityKey"/> that was used to sign this token.
         /// </summary>
-        /// <remarks>If the 'zip' claim is not found, an empty string is returned.</remarks>   
-        public string Zip => Header.Value<string>(JwtHeaderParameterNames.Zip) ?? String.Empty;
+        /// <remarks>
+        /// If the JWT was not signed or validated, this value will be null.
+        /// </remarks>
+        public override SecurityKey SigningKey { get; set; }
 
-        /// <summary>
-        /// Decodes the string into the header, payload and signature.
-        /// </summary>
-        /// <param name="tokenParts">the tokenized string.</param>
-        /// <param name="rawData">the original token.</param>
-        private void Decode(string[] tokenParts, string rawData)
+        internal byte[] MessageBytes{ get; set; }
+
+        internal int NumberOfDots { get; set; }
+
+        private void ReadToken(string encodedJson)
         {
-            try
+            // JWT must have 2 dots
+            Dot1 = encodedJson.IndexOf('.');
+            if (Dot1 == -1)
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+            if (Dot1 == encodedJson.Length)
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
+
+            Dot2 = encodedJson.IndexOf('.', Dot1 + 1);
+            if (Dot2 == -1)
+                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
```

Comment:
consider including a hint to differentiate these if PII isn't enabled. Currently we'd need someone to enable PII to differentiate from logs why their token is invalid which could delay investigations.

Reply:
@TimHannMSFT this should be PII as the token is logged.
I will make a change to strip off the signature.

Reply:
I was making a slightly different point I think. I agree the encoded json here should be scrubbed for PII, even without the signature the claims will often contain pii so need to be scrubbed.

My point was that if we see IDX14100 in the logs without PII scrubbing off, we have no way of knowing what was wrong with the token formatting. Maybe this is fine but I think it offers some diagnostic value to know that it didn't have any dots vs it was ended with the only period, vs it only had one dot but other text (and likewise, since we use the same error code for JWEs we can't distinguish those either).

It's possible this wouldn't help but I think it's worth the small work to add a suffix since it might save a back and forth with a customer getting logs. to narrow down some error case. Your call though, not something I'd block on.